File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章外卖店优先级.cpp
//外卖店优先级 2019年第十届蓝桥省赛C/C++大学A组
//lanqiaoOJ  184 

#include<bits/stdc++.h>
using namespace std;
const int N = 100010;
int order[N];  //order[id]  第id号店上一次的订单
int prior[N];  //prior[id]  第id号店的优先级
int flag[N];   //flag[id]   第id号店在不在优先缓存中
struct node{int time,id;}a[N];
bool cmp(node a, node b){  //结构体排序
    if(a.id == b.id) return a.time<b.time;
    return a.id<b.id;
}
int main(){
    int m,n,T;  cin>>n>>m>>T;
    for(int i=0;i<m;i++) cin>>a[i].time>>a[i].id;
    sort(a,a+m,cmp);           //按结构体中的时间排序
    for(int i=0;i<m;i++){
        int tt=a[i].time,  id=a[i].id;
        if(tt != order[id]) //如果当前订单不等于上一次的订单，减去它们之间的间隔
            prior[id] -= tt-order[id]-1;
        prior[id] = prior[id] < 0 ? 0: prior[id];  //不小于0
        if(prior[id]<=3)  flag[id]=0;
        prior[id]+=2;
        if(prior[id]> 5)  flag[id]=1;
        order[id] = tt;
    }
    for(int i=1;i<=n;i++)  //最后处理第T时刻
        if(order[i]<T){
            prior[i] -= T-order[i];
            if(prior[i]<=3)
                flag[i]=0;
        }
    int ans=0;
    for(int i=0;i<=n;i++)
        if(flag[i]) ans++;
    cout<<ans;
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章外卖店优先级.py
#外卖店优先级 2019年第十届蓝桥省赛C/C++大学A组
#lanqiaoOJ 184

n, m, T = map(int,input().split())
a = [] 
priorty = []
for i in range(m):
    a.append([int(i) for i in input().split()])
a = sorted(a, key=lambda a: a[0])  #按结构体中的时间排序
order = [0 for i in range(n+1)]
prior = [0 for i in range(n+1)]
flag = [0 for i in range(n+1)]
for i in range(m):
    tt = a[i][0]     #time
    idd = a[i][1]    #id
    if tt != order[idd]:
         prior[idd] -= tt-order[idd]-1
    if prior[idd] < 0:   prior[idd] =0
    if(prior[idd]<=3):   flag[idd]=0
    prior[idd] += 2
    if(prior[idd]> 5):   flag[idd]=1
    order[idd]=tt
for i in range (1,n+1):
    if(order[i]<T):
        prior[i] -= T-order[i]
        if(prior[i]<=3): flag[i]=0
ans=0
for i in range(n+1):
    if(flag[i]>0):
        ans += 1
print(ans)
